docs/csharp/pattern-matching.md

```diff
@@ -2,31 +2,192 @@
+virtual methods and overriden methods to customize object behavior based
+on the runtime type of the object.
 
-<< Build the sample in an additive way >>
-<< Do the shape thing for the https://docs.microsoft.com/en-us/dotnet/articles/fsharp/language-reference/discriminated-unions shapes >>
+Those techniques aren't possible for data that isn't structured in a class
+hierarchy. In those cases, you'd need to utilize traditional procedural
+imperative constructs like `if` or `switch` statements to manage the control
+flow.
 
--->
\ No newline at end of file
+The new *pattern matching* expressions enable cleaner syntax to examine data
+and manipulate control flow based on any conditon of that data.
+
+In this topic, you'll build a sample class that computes the area of
+different geometric shapes. But, you'll do it without resorting to object
+oriented techniques and building a class hierarchy for the different shapes.
+You'll use *pattern matching* instead. To further emphasize that we're not
+using inheritance, you'll make each shape a `struct` instead of a class. 
+
+## Using `is` pattern expressions ##
+
+Let's start simple with a square. 
+
+[!code-csharp[SquareDefinition](../../samples/csharp/PatternMatching/PatternMatching/Shapes.cs#01_SquareDefinition "Square definition")]
+
+The method to compute the area multiplies the length of a side by itself:
+
+[!code-csharp[02_ComputeSquareArea](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#02_ComputeSquareArea "Compute square area")]
+
+Next, we'll add a circle:
+
+[!code-csharp[CircleDefinition](../../samples/csharp/PatternMatching/PatternMatching/Shapes.cs#03_CircleDefinition "Circle definition")]
+
+We'll also write a method to compute the area of a circle:
+
+[!code-csharp[04_ComputeCircleArea](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#04_ComputeCircleArea "Compute circle area")]
+
+Ideally, you'd want one method that computes the area of any kind of
+shape. You've only created a couple different kinds of shapes so far.
+Let's use an `is` pattern matching expression to test the type of 
+shape:
+
+[!code-csharp[05_ComputeWithAsExpression](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#05_ComputeWithAsExpression "Compute with as expression")]
+
+The parameter type for this method is `object`: any type can be passed
+to this method. The code in the method checks the type to see if the parameter
+is either a square or a circle. Notice that the pattern match `is` expression
+can assign a variable as well as test the type. The pattern match expression
+`shape is Square s` is semantically the same as the more verbose syntax:
+
+```csharp
+Square s;
+if (shape is Square)
+{
+    s = (shape as Square);
+    // use 's'
+} 
+```
+
+Language rules for pattern matching expressions help you avoid misuising
+the results of a match expression. In the example above, the variables `s`
```

Comment:
and what about variable s1? you show it in the code but doesn't talk about it anywhere.